### ISSUE TITLE
Updating user code after the breaking Enter changes in 0.11

### ISSUE THREAD

**Issue by nmk** [haskell-servant/servant#810]
I am trying to update from 0.10 to 0.11. I found #734 and #736 with regards to the `enter` changes, but cannot see how to change user code to reflect those changes. This compiles with 0.10 and fails with 0.11:
```
data Config = Config
data Item = Item

type App = ReaderT Config Handler

getItemById :: Integer -> App Item
getItemById = undefined

type ItemApi = "item" :> Capture "itemId" Integer :> Get '[JSON] Item

server :: Config -> Server ItemApi
server config = enter (runReaderTNat config) getItemById
```
The error message with 0.11 is
```
    • Couldn't match type ‘Servant.Utils.Enter.Entered
                             (ReaderT Config n0) n0 (ReaderT Config Handler Item)’
                     with ‘Handler Item’
        arising from a use of ‘enter’
      The type variable ‘n0’ is ambiguous
    • In the expression: enter (runReaderTNat config) getItemById
      In an equation for ‘server’:
          server config = enter (runReaderTNat config) getItemById
   |
26 | server config = enter (runReaderTNat config) getItemById
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
How should the above example look for 0.11?

**Comment by phadej** [comment]
`The type variable ‘n0’ is ambiguous` means you need to add a type annotation.

```haskell
enter (runReaderTNat config :: App :~> Handler) getItemById
```

**Comment by nmk** [comment]
Thank you! Why is the type annotation needed where it wasn't previously?

**Comment by phadej** [comment]
In order to fix #736, one need to tell GHC either the shape (type) of `Server API` or the type of natural transformation. Otherwise there're ambiguous cases (e.g. when the source monad is `(->)`!). `enter` in `0.11`chose to require type for NT. in `0.12` there will be `hoistServer`, which fixes the shape of `Server`, see #804. That should be more ergonomic, yet we have to update tutorial (and add tests).

**Comment by nmk** [comment]
Thank you for the explanation!
